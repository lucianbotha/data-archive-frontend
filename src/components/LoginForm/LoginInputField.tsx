import * as React from "react";
import InputField from "../basicComponents/InputField";

/*
The login input field responsible for user input.

Properties:
-----------
label
  The label which detail the user input field
name
  The name that uniquecly identifies the input field
onChange
  The fuction that is responsible for updating the input field value when changed
placeholder
  It gives an example of the required value
type
  The type of the input field e.g. text, password, etc.
value
  The value of the input field
*/

interface ILoginInputProps {
  error?: string;
  label?: string;
  loading?: boolean;
  name?: string;
  onChange?: any;
  placeholder?: string;
  type?: string;
  value?: string;
}

const LoginInputField = (props: ILoginInputProps) => {
  const { error, label, loading, name, onChange, type, value } = props;
  return (
    <div className={"field"}>
      <label htmlFor={`${name}+${label}`} className={"label"}>
        {label}:
      </label>
      <div className={"control is-child"}>
        <InputField
          data-test="target-name-input"
          disabled={loading}
          name={name}
          value={value || ""}
          error={error}
          onChange={onChange}
          type={type}
        />
      </div>
    </div>
  );
};

export default LoginInputField;
