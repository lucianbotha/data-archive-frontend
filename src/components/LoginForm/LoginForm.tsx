import PropTypes from "prop-types";
import * as React from "react";
import styled from "styled-components";
import api from "../../api/api";
import { validateLoginField } from "../../util/LoginFormValidation";
import LoginInputField from "./LoginInputField";

/*
The login form which is responsible for authenticating th user into the data acrchive system.

Properties:
-----------
userInput : {
  password
    The password, which must have at least 7 characters.
  username
    The username, which must not contain upper case letters.
}

errors : []
  An array of error messages if any.
*/

interface ILoginFormState {
  errors: {
    password: string;
    responseError?: string;
    username: string;
  };
  userInput: {
    password: string;
    username: string;
  };
}

const LoginFormParent = styled.form.attrs({
  className: "column is-4 is-offset-4"
})`
  && {
    padding: 1px;
  }
`;

const Heading = styled.h1.attrs({
  className: "title is-3"
})`
  && {
    text-align: center;
    margin: 20px 0 20px 0;
  }
`;

const ErrorMessage = styled.p.attrs({
  className: "error tile"
})`
  && {
    text-align: left;
    margin: 3px 0 3px 0;
    padding: 2px 0 2px 0;
    background-color: hsl(348, 100%, 61%);
    color: white;
  }
`;

class LoginForm extends React.Component<ILoginFormState> {
  static contextTypes = {
    router: PropTypes.object
  };

  public state = {
    errors: {
      password: "",
      responseError: "",
      username: ""
    },
    loading: false,
    userInput: {
      password: "",
      username: ""
    }
  };

  onHandleSubmit = async (e: React.FormEvent<EventTarget>) => {
    e.preventDefault();

    // Validates the user input fields
    const errors = validateLoginField(this.state.userInput);

    // Update the errors propetry of the state
    this.setState({ errors });

    // Check if there is an error, if there is abort signing in.
    if (errors.password || errors.username) {
      return;
    }

    try {
      this.setState({
        loading: true
      });

      const login = await api.auth.login({
        ...this.state.userInput
      });

      if (login.data.success) {
        // TO BE REMOVED
        // Confirming user is authenticated
        const user = await api.user.queryUser();

        console.log(user);

        this.setState({
          errors: {
            password: "",
            responseError: "",
            username: ""
          },
          loading: false,
          userInput: {
            password: "",
            username: ""
          }
        });

        // Redirect the user to the home page granted access to his own content
        this.context.router.history.push("/");
      }
    } catch (error) {
      this.setState({
        errors: {
          ...this.state.errors,
          responseError: error.message
        },
        loading: false
      });
    }
  };

  onHandleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const name = e.target.name;
    const value = e.target.value;
    // Updating the userInput property of the state when input field value updates
    this.setState({
      userInput: {
        ...this.state.userInput,
        [name]: value
      }
    });
  };

  render() {
    const { errors, loading } = this.state;
    const { password, username } = this.state.userInput;
    return (
      <LoginFormParent onSubmit={e => this.onHandleSubmit(e)}>
        <Heading>Login Here</Heading>
        {errors.responseError ? (
          <ErrorMessage>{errors.responseError}</ErrorMessage>
        ) : null}
        <LoginInputField
          error={errors.username}
          loading={loading}
          name={"username"}
          label={"Username"}
          onChange={this.onHandleInputChange}
          type={"text"}
          value={username}
        />

        <LoginInputField
          error={errors.password}
          loading={loading}
          name={"password"}
          label={"Password"}
          onChange={this.onHandleInputChange}
          type={"password"}
          value={password}
        />

        <button
          disabled={loading}
          className="signIn button is-success is-fullwidth is-rounded"
        >
          {loading ? "Signing in..." : "Sign in"}
        </button>
      </LoginFormParent>
    );
  }
}

export default LoginForm;
