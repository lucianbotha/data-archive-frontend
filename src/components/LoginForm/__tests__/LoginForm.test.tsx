import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import * as React from "react";
import LoginForm from "../LoginForm";

// Helper function for simulating input field value change.
function inputTyping(wrapper: any, name: string, value: string) {
  wrapper.find(`input[name="${name}"]`).simulate("change", {
    target: { name, value }
  });
}

// Initial userInput state
const initialState = {
  errors: [],
  userInput: {
    password: "",
    username: ""
  }
};

// Updated userInput state
const updatedState = {
  errors: [],
  userInput: {
    password: "securepassword",
    username: "sj"
  }
};

describe("LoginForm Component", () => {
  it("renders the LoginForm component having unpopulated props with no errors", async () => {
    // RegistrationForm component wrapper.
    const wrapper = mount(
      <LoginForm
        userInput={initialState.userInput}
        errors={initialState.errors}
      />
    );
    // Expect the snapshot to match the RegistrationForm component.
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  // TODO
});
