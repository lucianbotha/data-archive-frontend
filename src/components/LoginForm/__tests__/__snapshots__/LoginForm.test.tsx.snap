// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm Component renders the LoginForm component having unpopulated props with no errors 1`] = `
<LoginForm
  errors={
    Object {
      "password": "",
      "username": "",
    }
  }
  userInput={
    Object {
      "password": "",
      "username": "",
    }
  }
>
  <styled.form
    onSubmit={[Function]}
  >
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [
            Object {
              "className": "column is-4 is-offset-4",
            },
          ],
          "componentStyle": ComponentStyle {
            "componentId": "sc-bwzfXH",
            "isStatic": true,
            "lastClassName": "fDYKEV",
            "rules": Array [
              "
  && {
    padding: 1px;
  }
",
            ],
          },
          "displayName": "styled.form",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-bwzfXH",
          "target": "form",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
      onSubmit={[Function]}
    >
      <form
        className="sc-bwzfXH column is-4 is-offset-4 fDYKEV"
        onSubmit={[Function]}
      >
        <styled.h1>
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [
                  Object {
                    "className": "title is-3",
                  },
                ],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-htpNat",
                  "isStatic": true,
                  "lastClassName": "hjCtQn",
                  "rules": Array [
                    "
  && {
    text-align: center;
    margin: 20px 0 20px 0;
  }
",
                  ],
                },
                "displayName": "styled.h1",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-htpNat",
                "target": "h1",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
          >
            <h1
              className="sc-htpNat title is-3 hjCtQn"
            >
              Login Here
            </h1>
          </StyledComponent>
        </styled.h1>
        <LoginInputField
          error=""
          label="Username"
          loading={false}
          name="username"
          onChange={[Function]}
          type="text"
          value=""
        >
          <div
            className="field"
          >
            <label
              className="label"
              htmlFor="username+Username"
            >
              Username
              :
            </label>
            <div
              className="control is-child"
            >
              <InputField
                data-test="target-name-input"
                disabled={false}
                error=""
                name="username"
                onChange={[Function]}
                type="text"
                value=""
              >
                <div
                  className="control"
                >
                  <input
                    className="input undefined "
                    data-test="target-name-input"
                    disabled={false}
                    name="username"
                    onChange={[Function]}
                    type="text"
                    value=""
                  />
                </div>
              </InputField>
            </div>
          </div>
        </LoginInputField>
        <LoginInputField
          error=""
          label="Password"
          loading={false}
          name="password"
          onChange={[Function]}
          type="password"
          value=""
        >
          <div
            className="field"
          >
            <label
              className="label"
              htmlFor="password+Password"
            >
              Password
              :
            </label>
            <div
              className="control is-child"
            >
              <InputField
                data-test="target-name-input"
                disabled={false}
                error=""
                name="password"
                onChange={[Function]}
                type="password"
                value=""
              >
                <div
                  className="control"
                >
                  <input
                    className="input undefined "
                    data-test="target-name-input"
                    disabled={false}
                    name="password"
                    onChange={[Function]}
                    type="password"
                    value=""
                  />
                </div>
              </InputField>
            </div>
          </div>
        </LoginInputField>
        <button
          className="signIn button is-success is-fullwidth is-rounded"
          disabled={false}
        >
          Sign in
        </button>
      </form>
    </StyledComponent>
  </styled.form>
</LoginForm>
`;
